## Image Classification
SVM vs Logistic Regression

In [1]:
import multiprocessing
from sklearn.model_selection import GridSearchCV
from sklearn import svm

In [2]:
cpu_cnt = multiprocessing.cpu_count()
print(cpu_cnt)

8


In [3]:
params = [
    {
    'kernel':['linear','rbf','poly','sigmoid'],
    'C' : [0.1,0.2,0.4,0.5,1.0,2.0,5.0]
    }
]

In [4]:
svc = svm.SVC()
gs = GridSearchCV(estimator=svc,param_grid = params,scoring='accuracy',cv=5,n_jobs=cpu_cnt)

### Load the Dataset

In [5]:
from sklearn.datasets import load_digits

In [6]:
digits = load_digits()

In [7]:
X = digits.data
Y  = digits.target
print(X.shape)
print(Y.shape)

(1797, 64)
(1797,)


## Logistic Regression

In [8]:
from sklearn.model_selection import cross_val_score

In [9]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

c:\users\apoorv\appdata\local\programs\python\python38-32\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\users\apoorv\appdata\local\programs\python\python38-32\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.

0.9137650882079852

Logistic Regresion gives around 92% accuracy!

## Support Vector Machines

In [10]:
svc = svm.SVC()
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()


0.9632838130609718

Pretty Bad Result :(

## Let's try changing HyperParameters using Grid Search

In [11]:
gs.fit(X,Y)### Grid search on SVM Parameters!


GridSearchCV(cv=5, estimator=SVC(), n_jobs=8,
             param_grid=[{'C': [0.1, 0.2, 0.4, 0.5, 1.0, 2.0, 5.0],
                          'kernel': ['linear', 'rbf', 'poly', 'sigmoid']}],
             scoring='accuracy')

In [12]:
gs.best_estimator_


SVC(C=5.0)

In [13]:
gs.best_score_

0.9738502011761063

SVM Gives 97% Accuracy using C=0.1, and Polynomial Kernel of Degree 3